### DISCUSSION TITLE
odd result on subimage search in IM 6.9.11.60

### DISCUSSION THREAD

**Discussion by nbi1** [ImageMagick/ImageMagick#4185] [Digital Image Processing]
Trying to do a subimage search as follows:

compare -metric RMSE -subimage-search source.png subimage.png locations.png > & /dev/null
convert locations-1.png -threshold 73% txt:|grep FFFFFF|cut --delimiter=':' --fields=1  > matches_found

This produces a list of matches like:

563,119
1187,119
11,157
11,158
484,231
905,231
1206,231
724,270
725,270
1251,270

Ok for the most part, but there's clearly a glitch. 11,157 and 11,158 are not different instances of the subimage. Neither are 724,270 and 725,270. In both cases there's a one pixel difference resulting in a spurious duplicate. Does anyone know why this might be happening and how to fix it? Yes, I could filter this with some post processing code, but that would be a kludge. The clean way to address this is to keep it from happening in the first place.

**Comment by fmw42** [comment]
If on Unix-like system (Linux, Mac OSX, Windows 10 Unix), you can use my script, maxima, to limit matches to some minimum distance between them. See http://www.fmwconcepts.com/imagemagick/index.html for my scripts. Basically it draw a small black region at a peak so that it is not found before getting the next highest peak.

You use the second output image from compare that is the correlation surface image as input to the script.

**Reply by nbi1** [reply]
Thanks, maxima helps as it does eliminate duplicates, but it identifies some false peaks along the way. The complete subimage match list as in my original post is:
11+157
11+158
11+308
11+309
88+645
382+309
481+645
484+231
563+119
679+645
679+646
724+270
725+270
905+231
1029+607
1135+494
1135+495
1187+119
1206+231
1251+270
1511+607

'maxima -r 25 -n 20 locations-1.png' produces:
11,158
11,309
88,645
123,157 X
382,309
404,382 X
481,645
484,231
563,119
679,645
725,270
802,231 X
905,231
1029,607
1135,494
1187,119
1206,231
1251,270
1322,118 X
1511,607

The duplicates have indeed been eliminated (very good), but false positives have been introduced (marked as X - not good). Why is maxima identifying these peaks when the preceding subimage search did not? When I use Gimp to see what is at these offsets in the original image I see what is going on. The subimage search was looking for 
![image](https://user-images.githubusercontent.com/6698428/132961401-c691d7c9-59d2-4f16-8c4e-6ffdf8bf4427.png). The maxima run correctly identified these, but it also identified instances of 
![image](https://user-images.githubusercontent.com/6698428/132961439-19034cc7-22b9-4870-bd8c-80f00e4a6a93.png) which is incorrect. I can filter out these false positives using your "similar" script (thanks for all the great tools!), but the question remains why maxima identified these peaks when the subimage search did not. I am grateful though for having tools to cope with this. It seems that "similar" will let me filter the undesirable peaks (0.581038 similarity for a false peak - clearly not a match).

**Reply by fmw42** [reply]
You told it to find 20 locations that are above some threshold, but you did not provide a threshold. So it got the highest 20 in the image. It knows nothing about your list. Use the -t argument to set your same threshold of 73.

**Reply by nbi1** [reply]
I made some unwarranted assumptions about maxima and its parameters. Also there doesn't seem to be a convenient way of specifying "return *all* maxima at a given threshold". If only t is used it stops after 1. If n is also used then one needs to know either the exact number of maxima determined previously somehow or make a guess at n so that it is likely bigger (as shown in the examples). Maybe a value of -1 could be used to indicate "*all* maxima at a given threshold". Further testing has shown that maxima does address my issue. Many thanks.

**Reply by fmw42** [reply]
Do you mean all maxima at or above that value? Did you try leaving off -n, but keep -t?

**Reply by fmw42** [reply]
Sorry, the default is 1 and a -n is require. I will have to look into that possibility.

**Reply by fmw42** [reply]
Try setting the -t and give -n a large number, say 100 or 1000. That should then stop at -t, but not do all 100 or 1000, especially if you know how many your threshold returned.

**Comment by snibgo** [comment]
> ... but there's clearly a glitch. 11,157 and 11,158 are not different instances of the subimage. 

The subimage matches, within your 73% threshold, at both locations. Whether you count these as genuinely different locations depends on your needs.

When I want, say, the ten best matches, I first find the best match. Then overpaint the input image at that location (as fmw42 suggests), then re-search to find the next match. Repeat until I have ten matches.

But I expect it can be done by searching just once, and finding peaks in the output from the search. Morphology can find peaks. See http://imagemagick.org/Usage/morphology/
